llama-cli/src/cli_args.rs

```diff
@@ -137,20 +170,192 @@ impl Args {
     }
 
     #[cfg(not(all(target_os = "macos", target_arch = "aarch64")))]
-    pub(crate) fn autodetect_num_threads(&self) -> usize {
+    pub fn autodetect_num_threads(&self) -> usize {
         num_cpus::get_physical()
     }
 
-    pub(crate) fn num_threads(&self) -> usize {
+    pub fn num_threads(&self) -> usize {
         self.num_threads
             .unwrap_or_else(|| self.autodetect_num_threads())
     }
-}
 
+    pub fn inference_session_parameters(&self) -> InferenceSessionParameters {
+        let mem_typ = if self.float16 {
+            ModelKVMemoryType::Float16
+        } else {
+            ModelKVMemoryType::Float32
+        };
+        InferenceSessionParameters {
+            memory_k_type: mem_typ,
+            memory_v_type: mem_typ,
+            repetition_penalty_last_n: self.repeat_last_n,
+        }
+    }
+
+    pub fn rng(&self) -> rand::rngs::StdRng {
+        if let Some(seed) = self.seed {
+            rand::rngs::StdRng::seed_from_u64(seed)
+        } else {
+            rand::rngs::StdRng::from_entropy()
+        }
+    }
+
+    pub fn inference_parameters(&self, session_loaded: bool) -> InferenceParameters {
+        InferenceParameters {
+            n_threads: self.num_threads(),
+            n_batch: self.batch_size,
+            top_k: self.top_k,
+            top_p: self.top_p,
+            repeat_penalty: self.repeat_penalty,
+            temperature: self.temperature,
+            bias_tokens: self.token_bias.clone().unwrap_or_else(|| {
+                if self.ignore_eos {
+                    TokenBias::new(vec![(EOT_TOKEN_ID, -1.0)])
+                } else {
+                    TokenBias::default()
+                }
+            }),
+            play_back_previous_tokens: session_loaded,
+            ..Default::default()
+        }
+    }
+}
 fn parse_bias(s: &str) -> Result<TokenBias, String> {
     s.parse()
 }
 
-/// CLI args are stored in a lazy static variable so they're accessible from
-/// everywhere. Arguments are parsed on first access.
-pub static CLI_ARGS: Lazy<Args> = Lazy::new(Args::parse);
+#[derive(Parser, Debug)]
+pub struct ModelLoad {
+    /// Where to load the model path from
+    #[arg(long, short = 'm')]
+    pub model_path: String,
+
+    /// A file to read the prompt from.
+    #[arg(long, short = 'f', default_value = None)]
+    pub prompt_file: Option<String>,
+
+    /// Sets the size of the context (in tokens). Allows feeding longer prompts.
+    /// Note that this affects memory. TODO: Unsure how large the limit is.
+    #[arg(long, default_value_t = 512)]
+    pub num_ctx_tokens: usize,
```

Comment:
By now I'm quite convinced the limit for this is 2048. Trying to generate beyond that value starts producing garbage. Setting a lower value can be used to slightly reduce memory usage, but has no effect on inference quality.

I don't want to make 2048 a hard error because we want to be able to run larger models if someone figures out a way to fine-tune LLaMA (or another model) to extend the context window. But we should document this is the maximum for LLaMA for the vast majority of users that don't know.

I think what we should do is set 2048 as the default, so that users don't need to worry about raising this. I think the increase in memory usage is not that much. Based on some napkin math, the full context window takes ~1.6GB of Ram (4096 embd * 2048 ctx * 24 layers * 4bytes), or half of it when using F16. The the default of 500 would be around ~400MB. But these are dwarfed anyway by the model's weights (especially since this value doesn't increase in larger models). So I don't think many people are benefitting from the lower default.